tuf/api/metadata.py

```
@@ -94,51 +90,68 @@ def as_json(self, compact: bool = False) -> None:
-
-    #                     break
-
-    #     return len(verified_keyids) >= key_ring.threshold.least
+    def sign(self, key: JsonDict, append: bool = False) -> JsonDict:
+        """Creates signature over 'signed' and assigns it to 'signatures'.
+
+        Arguments:
+            key: A securesystemslib-style private key object used for signing.
+            append: A boolean indicating if the signature should be appended
+                to the list of signatures or replace them.
+
+        Raises:
+            securesystemslib.exceptions.FormatError: Key argument is malformed.
+            securesystemslib.exceptions.CryptoError, \
+                    securesystemslib.exceptions.UnsupportedAlgorithmError:
+                Signing errors.
+
+        Returns:
+            A securesystemslib-style signature object.
+
+        """
+        signature = create_signature(key, self.signed.signed_bytes)
+
+        if append:
+            self.signatures.append(signature)
+        else:
+            self.signatures = [signature]
+
+        return signature
+
+    def verify(self, key: JsonDict) -> bool:
+        """Verifies 'signatures' over 'signed' that match the passed key by id.
+
+        Arguments:
+            key: A securesystemslib-style public key object.
+
+        Raises:
+            securesystemslib.exceptions.FormatError: Key argument is malformed.
+            securesystemslib.exceptions.CryptoError, \
+                    securesystemslib.exceptions.UnsupportedAlgorithmError:
+                Signing errors.
+
+        Returns:
+            A boolean indicating if all identified signatures are valid. False
+            if no signature was found for the keyid or any of the found
+            signatures is invalid.
+
+            FIXME: Is this behavior expected? An alternative approach would be
+            to raise an exception if no signature is found for the keyid,
+            and/or if more than one sigantures are found for the keyid.
```

```suggestion
            and/or if more than one signatures are found for the keyid.
```

---

I prefer the alternative approach of raising an exception if no signature is found, or more than one signature is found, for the keyid.

---

I think I do too. Any advice on the exception type? At some point we should discuss [exception taxonomy guidelines](https://github.com/theupdateframework/tuf/projects/1#card-44476165). But we shouldn't block on that.

---

We don't seem to have an exception in `tuf.exceptions` or `securesystemslib.exceptions` that would fit, we could either throw a generic exception _or_ define a new one. Either way, it will probably need some cleanup once guidelines are defined.